pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +55,83 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+type HeadersLazyLoaded struct {
+	Path   string
+	UserID string
+}
+
+type HeadersLazyLoadedTrackerState struct {
+	LazyLoadedBlocks map[string]int64 `json:"lazy_loaded_blocks"`
+	UserID           string           `json:"user_id"`
+}
+
+func (p *ReaderPool) persist(state HeadersLazyLoadedTrackerState, finalPath string) error {
+	// Create temporary path for fsync
+	tmpPath, err := os.CreateTemp("", "lazy-loaded")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpPath.Name())
+
+	data, err := json.MarshalIndent(state, "", "")
+	if err != nil {
+		return err
+	}
+
+	merr := multierror.New()
```

Comment:
It might be worth extracting this logic to an "atomic file create" method given we also use it for [the shutdown marker](https://github.com/grafana/mimir/blob/06ed798ad991e1939afd350ad7abbd9e9fedf2a1/pkg/util/shutdownmarker/shutdown_marker.go#L18)

Reply:
I might address this in a different PR and will see if it could be better the extract the common logic.